### PR TITLE
CRIMAPP-487 - Add residential property address page

### PR DESCRIPTION
**Pull request by ripan**
## Description of change
Add "What is the address of your client’s residential property?" page if client selects 'no' for the question "Is the address of the property the same as your client’s home address?" in the "Your client’s residential property" page

## Link to relevant ticket
[CRIMAPP-487](https://dsdmoj.atlassian.net/browse/CRIMAPP-487)

## Notes for reviewer
Currently we are routing to "which_savings_does_client_have" for all paths of capital assets. It will change as we introduce more pages to the journey

### Before changes:
Screenshot 2024-02-29 at 14 16 59

### After changes:
Screenshot 2024-02-29 at 14 14 24

## How to manually test the feature

[CRIMAPP-487]: https://dsdmoj.atlassian.net/browse/CRIMAPP-487?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/controllers/steps/capital/property_address_controller.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+      def property_record
+        @property_record ||= properties.find(params[:property_id])
+      rescue ActiveRecord::RecordNotFound
+        raise Errors::PropertyNotFound
+      end
+
+      def properties
+        @properties ||= current_crime_application.properties
+      end
```
Comment (mseedat-moj) [inline app/controllers/steps/capital/property_address_controller.rb:24]: Is there a need to memoize `properties` I think it is only used once on Line 17?

File: app/models/property.rb [hunk 2]
```diff
@@ -4,6 +4,10 @@ class Property < ApplicationRecord
   attribute :value, :pence
   attribute :outstanding_mortgage, :pence
 
+  store_accessor :address, :address_line_one, :address_line_two, :city, :country, :postcode
+
+  OPTIONAL_ADDRESS_ATTRIBUTES = %w[address_line_two].freeze
```
Comment (mseedat-moj) [inline app/models/property.rb:9]: There is `Address` DB table/model - is that inappropriate for property address?

**Review by mseedat-moj — COMMENTED** [review]
Some comments to help me understand implementation

File: app/models/property.rb (hunk 2 above)
Comment (ripan) [reply app/models/property.rb:9]: > There is `Address` DB table/model - is that inappropriate for property address?

During [Spike](https://dsdmoj.atlassian.net/wiki/spaces/CRIMAP/pages/4697555085/Capital+Land+journey+backend+spike) for ‘Land and Property’ discussion, we decided not to use Address table and make it polymorphic. We decided to use properties.address(json) column instead

File: config/locales/en/errors.yml [hunk 3]
```diff
@@ -510,6 +510,17 @@ en:
               inclusion: Select yes if the address of the property is the same as your client's home address
             has_other_owners:
               inclusion: Select yes if anyone else owns part of the land
+        steps/capital/property_address_form:
+          attributes:
+            address_line_one:
+              blank: Enter address line 1, typically the building and street
+            address_line_two: # optional
```
Comment (hiboabd) [inline config/locales/en/errors.yml:517]: If this is optional, does it need to be in the yml file?

File: db/migrate/20240229125141_change_column_address_in_properties.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+class ChangeColumnAddressInProperties < ActiveRecord::Migration[7.0]
+  def up
+    change_column(:properties, :address, :jsonb)
+  end
+
+  def down
+    change_column(:properties, :address, :json)
```
Comment (hiboabd) [inline db/migrate/20240229125141_change_column_address_in_properties.rb:7]: Why isn't this `:jsonb`?

Comment (ripan) [reply db/migrate/20240229125141_change_column_address_in_properties.rb:7]: @hiboabd I started with JSON and changed to JSONB just for the sake of consistency 😄

**Review by timpeat — APPROVED** [review]

File: app/services/decisions/capital_decision_tree.rb [hunk 5]
```diff
@@ -42,5 +44,12 @@ def after_property_type(property)
 
       edit(:properties, property_id: property)
     end
+
+    def after_properties
+      return edit(:property_address) if form_object.is_home_address.no?
```
[the remark this answers is not in the document]
Comment (ripan) [reply app/services/decisions/capital_decision_tree.rb:49]: @hiboabd Thanks for finding this issue. Awesome 😄 . It fixed now

**Review by hiboabd — APPROVED** [review]

**Pull request merged**